### ISSUE TITLE
I still get links errors despite #4978 landing

### ISSUE THREAD

**Issue by Eh2406** [rust-lang/cargo#5289]
I maintain a `-sys` crate but my users keep getting 
```
error: multiple packages link to native library `z`, but a native library can be linked only once
```
I thought that had been fixed in #4978 but it doesn't seem to be working. What can I do?

--------------------

In fact this is mostly a place for me to link people to, to unify the discussion. I will try to keep this up to date with any discussion below.

> it doesn't seem to be working

Here is why #4978 doesn't seem to be working for you.
#4978 only stops that error for crate with the `links` attribute in the [index](https://github.com/rust-lang/crates.io-index). That attribute only gets into the index if you publish the crate with a recent nightly/beta. So no protection for the older versions of your crate.

> What can I do?

You can manually make a PR to the [index](https://github.com/rust-lang/crates.io-index) to add the attribute to older versions.

Someone could scan all crates to find all the missing `links` attributes and add them.

It is not yet clear what the cargo team recommends at this time.

**Comment by klausi** [comment]
First prototype: https://github.com/klausi/cargo_fix_links/blob/master/src/main.rs
I have hard-coded everything to get the update for libz-sys working. Let's see if that pull request gets accepted and then we can refine the script from there.

**Comment by alexcrichton** [comment]
I'm personally not super enthusiastic about retroactively modifying the index in that I'm hoping that this issue will eventually solve itself as newer versions of crates are published. I think we may first want to send out a post recommending publishing new versions (and perhaps a tool detecting the crate you need to publish with newer versions of Cargo). After that's baked for a bit if it's still a big problem I think we could consider changing the index, but until that point we may want to hold off just yet

**Comment by Eh2406** [comment]
"Just wait for it to solve it self" Is a reasonable decision for what "the cargo team recommends". It has the disadvantage of slowing the time before `-Z minimal-versions` can be used effectively. As adding a new version does not resolve the issue of old versions not having the attribute. Eventually all the versions of the library that you are compatible with will be published with a new enough cargo, but that could be years if the library does not have a lot of breaking changes and you don't want to bump version just for the the attribute being in the index.

**Comment by alexcrichton** [comment]
Nah yeah I definitely agree that if we really want to push on `-Z minimal-version` we'd need to prioritize fixing this, probably through an index update. At this point though I don't think we're necessarily trying to get that flag to work too super smoothly, though.

**Comment by klausi** [comment]
OK, so then my alternative plan is to maintain a fork of the crates.io index at https://github.com/klausi/crates.io-index that has the `links` attributes fixed where necessary. That way I can now test compilation of minimal dependency versions with this:

```
CARGO_REGISTRY_INDEX=https://github.com/klausi/crates.io-index cargo build -Z minimal-versions
```

That at least helps with fixing crates manually such as https://github.com/alexcrichton/git2-rs/pull/307 . I think this is not super ideal to be added to CI configs, but a starting point that people can experiment with.

**Comment by alexcrichton** [comment]
Sounds like a plausible solution to me though! (albeit definitely transitionary)

**Comment by epage** [comment]
@Turbo87 have you all done a `links` backfill of the index?

**Comment by Turbo87** [comment]
not that I am aware of. do we have crates with differences between manifest and index?

**Comment by epage** [comment]
@Turbo87 from Eh2406's description, it sounds like this attribute existed in packages and was later added to the index and we were telling people to publish with newer cargo or manually backfill if they needed it.  This was 5 years ago, so likely people have published new versions since then or manually backfilled, so it would be a relatively low priority to do an automated backfill.

I'll transfer this issue over to the crates.io repo for you to decide whether its worth fixing or to close as "good enough".
